webserver/static-content/src/main/java/io/helidon/webserver/staticcontent/FileBasedContentHandler.java

```diff
@@ -0,0 +1,181 @@
+        return result;
+    }
+
+    /**
+     * Determines and set a Content-Type header based on filename extension.
+     *
+     * @param filename        a filename
+     * @param requestHeaders  an HTTP request headers
+     * @param responseHeaders an HTTP response headers
+     */
+    void processContentType(String filename,
+                            RequestHeaders requestHeaders,
+                            ResponseHeaders responseHeaders) {
+        // Try to get Content-Type
+        responseHeaders.contentType(detectType(filename, requestHeaders));
+    }
+
+    private MediaType detectType(String fileName, RequestHeaders requestHeaders) {
+        Objects.requireNonNull(fileName);
+        Objects.requireNonNull(requestHeaders);
+
+        // first try to see if we have an override
+        // then find if we have a detected type
+        // then check the type is accepted by the request
+        return findCustomMediaType(fileName)
+                .or(() -> MediaTypes.detectType(fileName)
+                        .map(MediaType::parse))
+                .map(it -> {
+                    if (requestHeaders.isAccepted(it)) {
+                        return it;
+                    }
+                    throw new HttpException("Media type " + it + " is not accepted by request",
+                                            Http.Status.UNSUPPORTED_MEDIA_TYPE_415);
+                })
+                .orElseGet(() -> {
+                    List<MediaType> acceptedTypes = requestHeaders.acceptedTypes();
+                    if (acceptedTypes.isEmpty()) {
+                        return MediaType.APPLICATION_OCTET_STREAM;
+                    } else {
+                        return acceptedTypes.iterator().next();
+                    }
+                });
+    }
+
+    Optional<MediaType> findCustomMediaType(String fileName) {
+        int ind = fileName.lastIndexOf('.');
+
+        if (ind < 0) {
+            return Optional.empty();
+        }
+
+        String fileSuffix = fileName.substring(ind + 1);
+
+        return Optional.ofNullable(customMediaTypes.get(fileSuffix));
+    }
+
+    void sendFile(Http.RequestMethod method,
+                  Path pathParam,
+                  ServerRequest request,
+                  ServerResponse response,
+                  String welcomePage)
+            throws IOException {
+
+        LOGGER.fine(() -> "Sending static content from file: " + pathParam);
+
+        Path path = pathParam;
+        // we know the file exists, though it may be a directory
+        //First doHandle a directory case
+        if (Files.isDirectory(path)) {
+            String rawFullPath = request.uri().getRawPath();
+            if (rawFullPath.endsWith("/")) {
+                // Try to found welcome file
+                path = resolveWelcomeFile(path, welcomePage);
+            } else {
+                // Or redirect to slash ended
+                redirect(response, rawFullPath + "/");
```

Comment:
This drops query params on redirect (same for `ClassPathContentHandler`). A request to `/fallback/existing?a=b` should redirect to `/fallback/existing/?a=b`. It would further be nice to make it configurable if the canonical URL that is redirected to has a trailing slash or not.

BTW, thanks for doing this @tomas-langer ! :+1:

Reply:
This code is copied over, and should be fixed a separate issue / PR.

Reply:
I have fixed it for this PR, so we have this working ;)